### PR TITLE
UsernamePassword and Authorization-code providers

### PR DESCRIPTION
**Pull request by NakulSabharwal**
- Authentication providers for UsernamePassword and AuthorizationCode.
- Tests for UsernamePassword and AuthorizationCode
- Added gitignore

### CODE REVIEW

File: src/main/java/com/microsoft/graph/auth/BaseAuthentication.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	}};
+
+	protected static String GetAuthority(NationalCloud authorityEndpoint, String tenant)
+	{
+		return CloudList.get(authorityEndpoint.toString()) + tenant;
+	}
+
+	protected String getScopesAsString() {
+		String scopeString = new String();
+		for(String s : this.Scopes) {
```
Comment (deepak2016) [inline src/main/java/com/microsoft/graph/auth/BaseAuthentication.java:60]: you may want to use StringBuilder to construct string.

File: src/main/java/com/microsoft/graph/auth/BaseAuthentication.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+						.setScope(getScopesAsString())
+						.setRefreshToken(response.getRefreshToken())
+						.setGrantType(GrantType.REFRESH_TOKEN)
+						.setScope(getScopesAsString())
+						.setRedirectURI(redirectUri);
+				if(this.ClientSecret != null) {
+					token.setClientSecret(this.ClientSecret);
+				}
+
+				OAuthClientRequest r = token.buildBodyMessage();
```
Comment (deepak2016) [inline src/main/java/com/microsoft/graph/auth/BaseAuthentication.java]: give meaningful name to variable r => request?

File: src/main/java/com/microsoft/graph/auth/confidentialClient/AuthorizationCodeProvider.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			List<String> scopes,
+			String authorizationCode,
+			String redirectUri,
+			NationalCloud nationalCloud,
+			String tenant,
+			String clientSecret){
+		
+		super(scopes,
+				clientId,
+				GetAuthority(nationalCloud==null?NationalCloud.Global:nationalCloud, tenant == null ? AuthConstants.Tenants.Common : tenant),
```
Comment (deepak2016) [inline src/main/java/com/microsoft/graph/auth/confidentialClient/AuthorizationCodeProvider.java]: add space after ? and :

File: src/main/java/com/microsoft/graph/auth/publicClient/UsernamePasswordProvider.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	protected OAuthClientRequest getTokenRequestMessage() throws OAuthSystemException {
+		String tokenUrl = this.authority + "/oauth2/v2.0/token";
+		TokenRequestBuilder token = OAuthClientRequest.
+				tokenLocation(tokenUrl)
+				.setClientId(this.ClientId)
+				.setUsername(this.Username)
+				.setPassword(this.Password)
+				.setGrantType(GrantType.PASSWORD)
+				.setScope(getScopesAsString());
+		OAuthClientRequest req = token.buildBodyMessage();
```
Comment (deepak2016) [inline src/main/java/com/microsoft/graph/auth/publicClient/UsernamePasswordProvider.java]: req => request ?

File: src/main/java/com/microsoft/graph/auth/BaseAuthentication.java (hunk 1 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/auth/BaseAuthentication.java:60]: Resolved this comment by using StringBuilder

Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/auth/BaseAuthentication.java]: Resolved this by using StringBuilder.

File: src/main/java/com/microsoft/graph/auth/BaseAuthentication.java (hunk 2 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/auth/BaseAuthentication.java]: Resolved this by changing variable name to request.

File: src/main/java/com/microsoft/graph/auth/confidentialClient/AuthorizationCodeProvider.java (hunk 3 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/auth/confidentialClient/AuthorizationCodeProvider.java]: I have now added space after ? and :

File: src/main/java/com/microsoft/graph/auth/publicClient/UsernamePasswordProvider.java (hunk 4 above)
Comment (NakulSabharwal) [reply src/main/java/com/microsoft/graph/auth/publicClient/UsernamePasswordProvider.java]: Resolved by changing variable name to request.

**Review by deepak2016 — APPROVED** [review]

**Review by dadheech115 — APPROVED** [review]

**Pull request merged**